--- 
title: "GIS and Spatial Analysis for Conservation Biology Using R"
author: "Dr. Yolanda Wiersma, Memorial University"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook #bookdown::html_document
documentclass: book
bibliography: [book.bib, packages.bib]
biblio-style: apalike
link-citations: yes 
git-hub-repo: nlboreal/ConsBiol_SpatialAnalysis
description: "Quantitative R skills for GIS and Spatial Analysis"
---




<!--chapter:end:index.Rmd-->

# Introduction {#intro}

Many biology-relevant skills are now performed using computers: statistical analyses, mathematical modelling, managing and reformatting data, visualizing data, identifying genes from DNA sequences, constructing 3D models of proteins, and making phylogenies. Quantitative skills and best practices in data science are now being incorporated throughout the biology major.

In Ecology and Conservation Biology, it is common for researchers to use spatial mapping tools such as Geographic Information Systems (GIS). The most common GIS software is the ESRI suite of products (ArcGIS, ArcPro, ArcONline). This software is expensive, and can be time consuming to learn. While R is not a GIS SYSTEM, it has the capabilities to create maps and to do many of tye types of spatial analyses that ecologists would use a GIS for. R can also be time-consuming to learn, but because R is free, we feel it will have advantages for students in that they will always be able to access it. Many ENGOs may wish to create maps and do spatial analysis, but lack the financial resources to purchase GIS. Thus, graduates who can do analysis in R will be well-positioned to contribute usefully in many conservation sectors.

<!--chapter:end:01-introduction.Rmd-->

# How to use this manual {#howto}

This manual has been developed specifically for Biology students. We are increasing the focus on quantitative skills in our Biology curriculum, in response to changing demands and emphasis within the field of biology itself. In 2020, we introduce first year biology (BIOL 1001/1002) students to R software and basic statistical concepts. Other courses (BIOL 2600 - Principles of Ecology, BIOL 3295 - Population and Evolutionary Ecology and BIOL 4607 - Models in Biology) all incorporate R tools in their labs. This course will teach you some R tools relevant to spatial analysis for conservation biology. 

If you are not familiar with R, it is recommended that you read through the MUN Biology R Manual, which is available online [here](https://ahurford.github.io/quant-guide-all-courses/), paying special attention to chapters 2-7.

While working through the manual, we recommend that you set up your computer screen so that you can see your browser and the R Studio software at the same time. You can experiment with re-sizing windows and if you have two monitors, that will be really handy. Or, if you have access to a laptop and a tablet, we recommend using RStudio on your laptop and opening the manual in the browser on your tablet so that you can move between the two easily. 

<!--chapter:end:02-HowToUseManual.Rmd-->

# What is GIS? {#whatIs}

GIS (Geographic Information Systems) is defined as "A geographic information system (GIS) is a framework for gathering, managing, and analyzing data. Rooted in the science of geography, GIS integrates many types of data. It analyzes spatial location and organizes layers of information into visualizations using maps and 3D scenes. With this unique capability, GIS reveals deeper insights into data, such as patterns, relationships, and situationsâ€”helping users make smarter decisions." (ESRI.com)

Many people think GIS is simply software for making maps. While you can make maps with a GIS, that is not the real purpose (to make really great maps, one would use cartographic software). A GIS is a spatial database. It allows you to store, organize, query, and analyze data - but not just any data - data in a GIS have specific spatial reference. Thus, the spatial location of data is linked to the attributes, and data can be analyzed based on both the spatial location AND the attributes of the data themselves. 

GIS is very powerful software and is used in many sectors of society. The biggest users of GIS software are in marketing, urban/municipal planning, and the military. Conservation and ecology researchers are actually in a small minority of GIS users.

This manual will introduce some of the commont things that conservation biologists will want to do with spatial data. These include:

1. visualizing data in a map
1. using sampling points to sample map data
1. using polygons to sample map data

As well, this manual will introduce some of the common issues and challenges when dealing with spatial data, including challenges with projections, map symbology, scale issues, and data organization. 

## Types of GIS Data

GIS data are divided into two categories, raster data and vector data. Another important component of GIS data is the metadata.

## Raster Data

Raster data represent the spatial data as a map made of pixels (sometimes called a "raster"). Each pixel has a unique value, in addition to a x, y coordinate. Raster data may be continuous, and represnet the actual value of the feature in the real world (e.g., soil pH) or raster data may be categorical, with numbers standing in for categorical data (e.g., 3 = decidous forest). In this case, the metadata should include a description of what the pixel data represent.

## Vector Data

Vector data represent the spatial data as either points, lines or polygons. Each feature in the dataset has georeferenced data (its locatoin along an x, y coordinate system) as well as attribute data. The attribute data can have many files. For example, a polygon feature representing protected areas might have fields for the name of the protected area, the year it was established, its size, the type of protected area (e.g., national, provincial or state park), and other data. Each line (or record) in the attribute data file is linked to a single feature in the data set. 

There are no rules for what kind of data must be represented as either points, lines or polygons, however there are some standard (and farily obvious) practices. Features that are generally linear in the real world (even if they are not perfectly one-dimensional), such as roads, rivers, and railways, are usually represented by lines. Sample points and single entities (e.g., single trees) are reprsented as points. Features that have 2 dimensions, such as parcels of land, are represented as polygons. Sometimes a feature may be shown as a point at one spatial scale (e.g., at the scale of a continent, cities are shown as points) but as polygons at another spatial scale (e.g., if we zoom in on the Northeast Avalon, we may want to show the municipal boundaries of the various cities and towns as polygons).

## Metadata

Metadata is "data about the data". Metadata is a very important component of good data practice and is discusssed in chapter 10 of the MUN Biology R manual. If you are unfamilair with the concept of metadata, you can review it [here](https://ahurford.github.io/quant-guide-all-courses/data.html). With GIS data, the metadata are especially important, as they contain information about the coordinate reference system, or CRS (see chapter \@ref(projections)), as well as the attribute data, along with information about how and when the data were created. 

<!--chapter:end:03-WhatIsGIS.Rmd-->

# Reading in Vector Data {#vector}

We will use the R package ```sf``` for handling vector data. You will have to download and install this package from the CRAN repository (review section [4.10](https://ahurford.github.io/quant-guide-all-courses/rintro.html#r-packages) of the MUN Biology R manual if need a reminder of how to do this).

```{r, message = FALSE}
library(sf)
```

To plot the data (i.e., make a map) we will also need the package ```ggplot2```, so make sure to load and install that as well.

```{r}
library(ggplot2)
```

```{r, echo = FALSE}
setwd("C:/Users/ywiersma/Documents/BIOL4651/GIS_R_manual/ConsBiol_SpatialAnalysis")
```

Download the folder "HARV" to your working directory. If you do not remember how to check, and set your working directory, review section [4.2](https://ahurford.github.io/quant-guide-all-courses/rintro.html#working-directory). This folder contains data from the Harvard Research Forest. The vector data include a point file that represents sampling points, a line file that is the boundary of the resarch forest, and two polygon files that show the boundary of the research forest and the soil regions. 

We read these files in in a similar way as you read in a file like a csv file, but instead of using ```read.csv``` we use ```st_read()```

Read in the boundary shapefile (recognizable by the .shp extension) for the Harvard forest

```{r}
HARV_boundary <- st_read("HARV/harv_boundary.shp")
```

Read in the the soil region shapefile for the Harvard forest

```{r}
HARV_soils <- st_read("HARV/harv_soils.shp")
```

To create a map we will use the ```ggplot()``` function, with the ```geom_sf``` function.

```{r}
ggplot() +
  geom_sf(data = HARV_boundary) + 
  geom_sf(data = HARV_soils)
```

GIS maps are all about "layers" of data. The order in which data are displayed in ```ggplot``` is important. In the above code, we "layered" the ```HARV_soils``` polygon *over* the ```HARV_boundary``` polygon. Check what happens when we layer them in the opposite order: 

```{r}
ggplot() +
  geom_sf(data = HARV_soils) + 
  geom_sf(data = HARV_boundary)
```

Now the boundary file is over top the soils one and because it is a filled polygon, we can't see the soils layer. One way we can deal with this, is by adjusting the transparency of the boundary layer with the ```alpha``` function. This adjusts the transparency from fully transparent (```alpha = 0```), to fully opaque (```alpha = 1```). 

```{r}
ggplot() +
  geom_sf(data = HARV_soils) + 
  geom_sf(data = HARV_boundary, alpha = 0.1)
```

We can also symbolize data based on the attributes. Before we do that, let's explore the HARV_soils layer. If you type ```HARV_soils``` At the ```>``` prompt, you will see the metadata. You will see that in addition to some descriptions of the geographic reference system (for more on that, see the chapter \@ref(projections)), you will see some attributes, including different kinds of soil classification and drainage types. If we want to display the soil polygons by their SIMMONS_SO field, we would use the ```aes``` (for aesthetics) function.

```{r}
ggplot() +
  geom_sf(data = HARV_soils, mapping = aes(fill = SIMMONS_SO)) + 
  geom_sf(data = HARV_boundary, alpha = 0.1)
```

The default colour scheme has been applied. You can change this using different built in colour maps. Below is the code to use the ```viridis``` colour map.

```{r}
ggplot() +
  geom_sf(data = HARV_soils, mapping = aes(fill = SIMMONS_SO)) + 
  geom_sf(data = HARV_boundary, alpha = 0.1) +
  scale_fill_viridis_d()
```

To see other aesthetics that can be modified, run ```vignette("ggplot2-specs")```


<!--chapter:end:04-Vectors.Rmd-->

# Reading in Raster Data {#raster}

We will use the R package ```stars``` for handling raster data. You will have to download and install this package from the CRAN repository (review section [4.10](https://ahurford.github.io/quant-guide-all-courses/rintro.html#r-packages) of the MUN Biology R manual if need a reminder of how to do this).

```{r, message= FALSE}
library(stars) 
```

To plot the data (i.e., make a map) we will also need the package ```ggplot2```, so make sure to load and install that as well, if you haven't yet already.

```{r}
library(ggplot2)
```

If you didn't already set the working directory to where you have downloaded the HARV folder, do so now. This folder contains 4 .tif files. You may be familiar with .tif files as digital photographs. These are a special kind of .tif called a geotiff. Like a digital photograph, it has pixels, but it also has georeferenced data. 

```{r, echo = FALSE}
#setwd("C:/Users/ywiersma/Documents/BIOL4651/GIS_R_manual/ConsBiol_SpatialAnalysis/HARV")
```

We will read in the raster that represents the digital terrain model (dtm for short) - in this raster, each pixel gives the elevation above sea level of that point. We use the ```read_stars``` function.

```{r}
HARV_dtm <- read_stars("HARV/HARV_dtmFull.tif")
```

To plot it, we use ```geom_stars()``` in ```ggplot()```. We'll use the ```viridis``` colour palette. Note that we specified _c (for continous) here, because the raster values are continuous data, whereas when we used the colour palette for the soils polygon data in the \@ref(vector) chapter, we used _d (for discrete), because those data were discrete categories.

```{r}
ggplot() +
  geom_stars(data = HARV_dtm) +
  scale_fill_viridis_c()
```

**NOTE**: If it takes too long to draw the map, try loading in a subset of the dtm (the "HARV_DTMCrop.tif" file) instead. Raster files can take a long time to load and depending on your computer's available CPU, you may wish to use a smaller raster layer)

What if we want to show the boundaries of the Harvard forest on top of this raster layer? Add it in ggplot like this:

```{r}
ggplot() +
  geom_stars(data = HARV_dtm) +
  geom_sf(data = HARV_boundary, alpha = 0.1) +
  scale_fill_viridis_c()
```


Let's try making another map that combines raster and vector data. Download the SJER folder, and change your working directory to that folder. This is similar data to that from the Harvard forest, but from the San Joaquin Experimental Range. On your own, read in the following layers and try to plot them.
1. sjer_plots (these are point files that represent the location of the sample plots)
1. SJER_dtmFULL.tif (a raster of the digital terrain model showing elevations)

```{r, echo = FALSE}
#setwd("C:/Users/ywiersma/Documents/BIOL4651/GIS_R_manual/ConsBiol_SpatialAnalysis/SJER")
```

```{r}
SJER_plots <- st_read("SJER/sjer_plots.shp")
SJER_dtm <- read_stars("SJER/SJER_dtmFULL.tif")

ggplot() +
  geom_stars(data = SJER_dtm) +
  geom_sf(data = SJER_plots)
```

You will get a map that does not quite look right. Congratulations - you've discovered the most common (and frustrating) issue with spatial data - conflicting projections. To learn more, go the the chapter on projections; chapter \@ref(projections). 

  

<!--chapter:end:05-Rasters.Rmd-->

# Dealing with Projections {#projections}

At the end of the vectors \@ref(vector) chapter, we tried to map the plots layer for the San Joaquin Experimental Range over the digital terrain model (dtm). The two layers didn't display properly, so something went wrong. To see what happened, we need to have a look at the co-ordinate reference system (crs) for the two layers. Use the ```st_crs``` command to see what this is for each layer.

```{r}
st_crs(SJER_plots)
st_crs(SJER_dtm)
```

Scroll through the results and compare the last set of numbers at the end of the file in the ID[] line. You should note that the SJER_plots file as an ID of 4325 and that the SJER_dtm has an ID of 32611. This indicates that the two layers have different coordinate systems. You can verify this by plotting each separately and cheeking the coordinate system labels on the map edges. The labels on the dtm are in UTM units, while the labels on the plots are in latitude and longitude. We need to convert the coordinate reference system (crs) of one of the layers to match the other. We use the ```st_transform``` function to create a new plots layer which has a projection system that matches the dtm layer.

```{r}
SJER_plots_utm <- st_transform(SJER_plots, st_crs(SJER_dtm))
```

Now you should be able to plot the two layers together

```{r}
ggplot() +
  geom_stars(data = SJER_dtm) +
  geom_sf(data = SJER_plots_utm)
```

Now that you have the two data sets aligned, you can extract the elevation data for each point, which is explained in the chapter on extracting data \@ref(extractingData). Before you do that, complete the exercises in the next chapter \@ref(rasterAlgebra).  

<!--chapter:end:06-Projections.Rmd-->

# Raster Algebra {#rasterAlgebra}

It is possible to analyse and combine two different rasters. The second raster layer in the SJER folder is a digital surface model (dsm) which is a data set taken using Lidar data that measures the height of the objects (above sea level) of objects on the earth. Let's load that layer.

```{r}
SJER_dsm <- read_stars("SJER/SJER_dsmCROP.tif")
```

By subtracting the dtm (elevation of the ground above sea level) from the dsm (elevation of the objects on the ground above sea level) we can create a new raster that represents the height of those objects above the ground. Since this layer is in a forest, we can infer that this new layer is the canopy height. However, the dsm layer covers a smaller extent that the dtm, so we'll also read in the cropped .tif file for the dtm

```{r}
SJER_dtm_cropped <- read_stars("SJER/SJER_dtmCROP.tif")
```

Now we can do the raster algebra: 

```{r}
SJER_canopy = SJER_dsm - SJER_dtm_cropped
```

Raster algebra can use any algebraic functions and involve one or more raster layers - for example if we wanted to convert the units of the canopy layer from metres to centimetres, we could simply multiply the SJER_canopy layer by 10. Or if you had a raster of canopy height in 2000 and canopy height in 2020, you could subtract one from the other to find out how much the tree height changed over 20 years in different parts of the landscape. 

Now we can extract the canopy height data from our sample locations - see chapter  \@ref(extractingData). 

<!--chapter:end:07-RasterAlgebra.Rmd-->

# Extracting Data from a Raster {#extractingData}

A common thing that ecologists and conservation biologists want to get data for their sample sites from different GIS layers and then use these data in different models or statistical tests.

## Sampling data from a raster using points

We use the ```aggregate``` function to get data from the canopy raster

```{r}
plot_canopy_height <- aggregate(SJER_canopy, SJER_plots_utm, mean, as_points = FALSE)
plot_canopy_height$SJER_dsmCrop.tif
```

To link this data into the attribute data we will use the ```mutate``` function in the package ```dplyr```. We have a lab in BIOL 1002 that teaches some of the basics of the dplyr package. If you're not familiar with that package, just use the code below.

```{r, message = FALSE}
library(dplyr)
```

```{r}
mutate(SJER_plots_utm, canopyHeight = plot_canopy_height$SJER_dsmCrop.tif)
```

If you examine the SJER_plots_utm data, you will see that it now has the canopy height data in the attribute data. You'll notice that there is also an attribute for ```plot_type```. We can make a map that symbolizes the sample points by the plot type like this:

```{r}
ggplot() +
  geom_stars(data = SJER_canopy) +
  geom_sf(data = SJER_plots_utm, mapping = aes(color = plot_type))
```

You could now take this data and test whether canopy height is significantly different at the tower sampling points or at the distributed sampling points. We won't do that just now; we'll do some statistics on these data in a different module. 


## Sampling data from a raster using polygons

At other times, researchers may want to sample data from a raster using a polygon. For example, we might want to know what the maximum elevation within the boundaries of a protected area is. 

To do these exercises, we're going to change back to Harvard forest data set. Set your working directory to the HARV folder.

```{r, echo = FALSE}
#setwd("C:/Users/ywiersma/Documents/BIOL4651/GIS_R_manual/ConsBiol_SpatialAnalysis/HARV")
```

We're going to find out what the maximum elevation is in each of the soil drainage polygons.

First, check that the two layers have the same co-ordinate reference system:

```{r}
st_crs(HARV_dtm)
st_crs(HARV_soils)
```

The ID code at the end of each file is identical (```32618```), so you can overlay these layers and extract the data. If you are not convinced, try mapping them together.

Then we use the same ```aggregate``` function as we did to extract the data from the raster to points. Note, that when you extract with a polygon, there are many pixels within the polygon. You can specify different values you want to extract from the pixels within the polygon, for example, maximum, minimum, mean. 

```{r}
drainage_max_elev <- aggregate(HARV_dtm, HARV_soils, max, as_points = FALSE)
drainage_max_elev$HARv_dtmFull.tif
```

Here we'll use the ```mutate``` function and create a new shapefile

```{r}
elevation_by_drainage <- mutate(HARV_soils, elevation = drainage_max_elev$HARV_dtmFull.tif)
```

Now we can plot the soil drainage basins and colour code them by their maximum height.

```{r}
ggplot() +
  geom_sf(data = elevation_by_drainage, mapping = aes(fill = elevation)) +
  scale_fill_viridis_c()
```

If you want to change the map display so the coordinates on the map boundary are in UTM, you can do it like so:

```{r}
ggplot() +
  geom_sf(data = elevation_by_drainage, mapping = aes(fill = elevation)) +
  coord_sf(datum = st_crs(HARV_dtm)) + 
  scale_fill_viridis_c()
```

What if you have sampled points in the field with a handheld gps and only have a csv file with x and y coordinates? How can you get the data? Go to chapter \@ref(PointsData) to find out! 

<!--chapter:end:08-ExtractingData.Rmd-->

# Creating a Points Shapefile from Data {#PointsData}

Often you will have a .csv file with x and y coordinates (either in lat/lon or UTM) that you sampled in the field and want to load into a GIS so that you can extract data from other GIS layers. Here's how to do it. We'll jump back to the San Joaquin Data set, so reset your working directory to the SJER folder (you're getting lots of practice with working directories this way!)

```{r, echo = FALSE}
#setwd("C:/Users/ywiersma/Documents/BIOL4651/GIS_R_manual/ConsBiol_SpatialAnalysis/SJER")
```

First we need to load the csv plots data as an ````sf``` object and display it

```{r}
plots_sjer_csv <- st_read("SJER/sjer_plots.csv", options = c("X_POSSIBLE_NAMES=longitude", "Y_POSSIBLE_NAMES=latitude"), crs = 4326)
```  
Then we'll re-project it to the UTM coordinates of the digital terrain models

```{r}
plots_sjer_csv_utm <- st_transform(plots_sjer_csv, st_crs(SJER_dtm))
```

Now you can proceed to map these points with any of the other SJER layers, as you did before, if you'd like to do more practice. 

<!--chapter:end:09-CreatePoints.Rmd-->

